src/cargo/ops/cargo_compile.rs

```diff
@@ -408,13 +408,24 @@ fn scrape_build_config(config: &Config,
 fn scrape_target_config(config: &Config, triple: &str)
                         -> CargoResult<ops::TargetConfig> {
 
+    let get_path_string = |key: &str| -> CargoResult<Option<String>> {
+        Ok(
+            if let Some(path) = try!(config.get_path(key)) {
+                Some(try!(path.into_os_string().into_string()
+                          .map_err(|_| internal("unhandled non-utf8 path"))))
+            } else {
+                None
+            }
+        )
+    };
+
     let key = format!("target.{}", triple);
-    let ar = try!(config.get_string(&format!("{}.ar", key)));
-    let linker = try!(config.get_string(&format!("{}.linker", key)));
+    let ar = try!(get_path_string(&format!("{}.ar", key)));
+    let linker = try!(get_path_string(&format!("{}.linker", key)));
 
     let mut ret = ops::TargetConfig {
-        ar: ar.map(|p| p.0),
-        linker: linker.map(|p| p.0),
+        ar: ar,
+        linker: linker,
```

Comment:
I initially attempted that, but backed off as it just moves the convert-to-utf8 problem from one place with convenient error reporting to another with far less convenient error reporting. That PathBuf ultimately ends up in a string, it's just a question of when.

Converting to utf8 within `fn scrape_target_config` of `src/cargo/ops/cargo_compile.rs` has the advantage that it already has error handling in the form of a `CargoResult` return type.

The alternative is to convert to utf8 at the point of use, within `fn build_plugin_args` of `src/cargo/ops/cargo_rustc/mod.rs`. The disadvantage of this is that there is no convenient way to handle the (unlikely but) potential error.

Options are:

1. keep `linker` and `ar` fields as `String`s within `TargetConfig`, or
1. change `linker` and `ar` fields to `PathBuf`s within `TargetConfig`, and handle the convert-to-utf8 by:
 1. weaving `CargoResult<()>` throughout the `fn build_plugin_args` call chain and failing gracefully upon utf8 conversion error, or by
 1. failing ungracefully upon utf8 conversion error.

The code thus far follows the path of least resistance for the best UX, option 1 above.

Let me know if you'd like me to pursue option 2i, 2ii or something else that I've probably overlooked :)